include/seqan3/range/views/kmer_hash.hpp

```diff
@@ -290,16 +281,66 @@ class kmer_hash_view<urng_t>::shape_iterator
     * Linear in size of shape.
     */
     shape_iterator(it_t it_start, sentinel_t it_end, shape s_) :
-        shape_{s_}, text_left{it_start}, text_right{std::ranges::next(it_start, s_.size(), it_end)}
+        shape_{s_}, text_left{it_start}, text_right{std::ranges::next(text_left, shape_.size() - 1, it_end)}
     {
         assert(std::ranges::size(shape_) > 0);
 
-        if (shape_.size() <= std::ranges::distance(text_left, text_right))
+        // shape size = 3
+        // Text:      1 2 3 4 5 6 7 8 9
+        // text_left: ^
+        // text_right:    ^
+        // distance(text_left, text_right) = 2
+        if (shape_.size() <= std::ranges::distance(text_left, text_right) + 1)
         {
             roll_factor = pow(sigma, static_cast<size_t>(std::ranges::size(shape_) - 1));
+            hash_full();
+        }
+    }
 
+    /*!\brief Construct from a given iterator on the text and a seqan3::shape.
+    * /param[in] it_start Iterator pointing to the first position of the text.
+    * /param[in] it_end   Sentinel pointing to the end of the text.
+    * /param[in] s_       The seqan3::shape that determines which positions participate in hashing.
+    * /param[in] is_end   Indicates that this iterator should point to the end of the text.
+    *
+    * \details
+    *
+    * If we have a common_range as underlying range, we want to preserve this property.
+    * This means that we need to have a shape_iterator that can act as end for the kmer_hash_view, i.e.
+    * the text_right iterator is equal to the end iterator of the underlying range.
+    * However, we still need to do some initialisation via hash_full:
+    * When using `std::views::reverse`, we start iterating from the end and decrement the iterator.
+    * After calling hash_full, we need to reset our text_right iterator to point to the end again.
+    *
+    * Another difference to the other constructor is that we need to do some work to determine the position of
+    * the text_left iterator. Note that we use `std::ranges::next` instead of `std::ranges::prev` because the latter
+    * only works for bidirectional ranges.    *
+    *
+    * ### Complexity
+    *
+    * Linear in size of shape.
+    */
+    shape_iterator(it_t it_start, sentinel_t it_end, shape s_, bool SEQAN3_DOXYGEN_ONLY(is_end)) : shape_{s_}
+    {
+        assert(std::ranges::size(shape_) > 0);
+
+        auto urange_size = std::ranges::distance(it_start, it_end);
+        auto step = (shape_.size() + 1 > urange_size) ? 0 : urange_size - shape_.size() + 1;
+        text_left = std::ranges::next(it_start, step, it_end);
```

Comment:
Mhmm, since we do this anyway, would it not be sufficient to use this in the end function and call the old constructor with text_it put to the respective position?

Reply:
If I got it correct, no, because then the right text iterator would point one before the end instead to the end

Reply:
Ok, so if I get it right, then the shape_iterator stores the begin and the end of the current infix over the text to compute the hash. That would mean text_lext points to the first valid character in the shape and text_right to the one behind the last one of the shape. Which is ok, because we ensure that after the last increment (when we reached the end) the text_right points to the regular end of the underlying range (in case it is a common range). In the other case we compare the text_right with the sentinel. So why does it point to one before the end?

Reply:
I have to look at it again, it was quite confusing to make it point to where it should.